plugins/woocommerce/src/StoreApi/Schemas/V1/CheckoutSchema.php

```diff
@@ -273,27 +277,81 @@ protected function get_additional_fields_response( \WC_Order $order ) {
 	 * @return array
 	 */
 	protected function get_additional_fields_schema() {
-		$additional_fields_keys = $this->additional_fields_controller->get_additional_fields_keys();
-
-		$fields = $this->additional_fields_controller->get_additional_fields();
-
-		$additional_fields = array_filter(
-			$fields,
-			function( $key ) use ( $additional_fields_keys ) {
-				return in_array( $key, $additional_fields_keys, true );
-			},
-			ARRAY_FILTER_USE_KEY
-		);
+		$order_only_fields = $this->additional_fields_controller->get_order_only_fields();
 
 		$schema = [];
-		foreach ( $additional_fields as $key => $field ) {
-			$schema[ $key ] = [
+		foreach ( $order_only_fields as $key => $field ) {
+			$field_schema = [
 				'description' => $field['label'],
 				'type'        => 'string',
 				'context'     => [ 'view', 'edit' ],
-				'required'    => true,
+				'required'    => $field['required'],
 			];
+
+			if ( 'select' === $field['type'] ) {
+				$field_schema['enum'] = array_map(
+					function( $option ) {
+						return $option['value'];
+					},
+					$field['options']
+				);
+			}
+
+			if ( 'checkbox' === $field['type'] ) {
+				$field_schema['type'] = 'boolean';
+			}
+
+			$schema[ $key ] = $field_schema;
 		}
 		return $schema;
 	}
+
+
+	/**
+	 * Sanitize and format additional fields object.
+	 *
+	 * @param array            $fields Values being sanitized.
+	 * @param \WP_REST_Request $request The Request.
+	 * @return array
+	 */
+	public function sanitize_additional_fields( $fields, $request ) {
+		$fields = array_reduce(
+			array_keys( $fields ),
+			function( $carry, $key ) use ( $fields, $request ) {
+				$carry[ $key ] = rest_sanitize_request_arg( wp_unslash( $fields[ $key ] ), $request, $key );
```

Comment:
This doesn't work correctly, if you look into `rest_sanitize_request_arg` it looks like this:

```php
function rest_sanitize_request_arg( $value, $request, $param ) {
	$attributes = $request->get_attributes();
	if ( ! isset( $attributes['args'][ $param ] ) || ! is_array( $attributes['args'][ $param ] ) ) {
		return $value;
	}
	$args = $attributes['args'][ $param ];

	return rest_sanitize_value_from_schema( $value, $args, $param );
}
```

The first condition is met so the original value is returned without being sanitized. Maybe you can use `rest_sanitize_value_from_schema`?

By the way, `rest_sanitize_value_from_schema` is called during `rest_validate_value_from_schema` which happens in `validate_additional_fields`. I am not sure if it needs to be done in the `sanitize_callback` too, before it gets to validation?

Reply:
Do I understand that replacing `rest_sanitize_request_arg` with `rest_sanitize_value_from_schema` would fix this? otherwise just skipping sanatize at all?

Reply:
> Do I understand that replacing rest_sanitize_request_arg with rest_sanitize_value_from_schema would fix this? otherwise just skipping sanatize at all?

Yes, but you need to make sure the correct values are being passed to `rest_sanitize_value_from_schema`, it's not just a straight swap.